### PR TITLE
Dtls multi core

### PR DESCRIPTION
**Pull request by JacobBarthelmeh**
Adds the option to use --enable-sessionexport with DTLS and send a session after completing change cipher spec in handshake.

### CODE REVIEW

**Comment by JacobBarthelmeh** [comment]
retest this please Jenkins

**Comment by JacobBarthelmeh** [comment]
Made the changes talked about in code review.

**Comment by toddouska** [comment]
On Mac I get some unused function errors on basic ./configure and ./configure --enable-dtls

src/internal.c:361:20: error: unused function 'c64toa'
      [-Werror,-Wunused-function]
static INLINE void c64toa(word64 u64, byte* c)
                   ^
src/internal.c:399:20: error: unused function 'ato64'
      [-Werror,-Wunused-function]
static INLINE void ato64(const byte* c, word64* u64)

File: wolfssl/ssl.h [hunk 1]
```diff
@@ -225,6 +225,20 @@ WOLFSSL_API WOLFSSL_METHOD *wolfTLSv1_2_client_method(void);
...
+#ifdef WOLFSSL_SESSION_EXPORT
+#ifdef WOLFSSL_DTLS
+typedef int (*wc_dtls_export)(WOLFSSL* ssl,
+                   unsigned char* exportBuffer, unsigned int sz, void* userCtx);
+WOLFSSL_API int wolfSSL_dtls_import(WOLFSSL* ssl, unsigned char* buf,
+                                                               unsigned int sz);
+WOLFSSL_API int wolfSSL_CTX_dtls_set_export(WOLFSSL_CTX* ctx,
+                                                           wc_dtls_export func);
+WOLFSSL_API int wolfSSL_dtls_set_export(WOLFSSL* ssl, wc_dtls_export func);
+WOLFSSL_API int wolfSSL_dtls_export(unsigned char* buf, unsigned int* sz,
```
Comment (toddouska) [inline wolfssl/ssl.h]: I think having WOLFSSL parameter first is more consistent.

File: src/ssl.c [hunk 2]
```diff
@@ -153,6 +153,129 @@ char* mystrnstr(const char* s1, const char* s2, unsigned int n)
...
+        return BAD_FUNC_ARG;
+    }
+
+    ssl->dtls_export = func;
+
+    return SSL_SUCCESS;
+}
+
+
+int wolfSSL_dtls_export(unsigned char* buf, unsigned int* sz, WOLFSSL* ssl)
```
Comment (toddouska) [inline src/ssl.c]: A comment about return codes on failure (SSL_FAILURE?) and how/why this differs from callback variant will make maintenance easier.

File: src/ssl.c [hunk 3]
```diff
@@ -153,6 +153,129 @@ char* mystrnstr(const char* s1, const char* s2, unsigned int n)
...
+    }
+
+    /* if not DTLS do nothing */
+    if (!ssl->options.dtls) {
+        WOLFSSL_MSG("Currently only DTLS export is supported");
+        return 0;
+    }
+
+    /* copy over keys, options, and dtls state struct */
+    ret = wolfSSL_dtls_export_internal(buf, *sz, ssl);
```
Comment (toddouska) [inline src/ssl.c]: If no conversions or messages are needed on return code the next two returns are redundant and simply returning the export_internal() is sufficient.

File: src/ssl.c [hunk 4]
```diff
@@ -153,6 +153,129 @@ char* mystrnstr(const char* s1, const char* s2, unsigned int n)
...
+    if (ret < 0) {
+        return ret;
+    }
+
+    return ret;
+}
+
+
+/* returns 0 on success */
```
Comment (toddouska) [inline src/ssl.c]: How/why this differs from manual export will make maintenance easier in the future.

**Comment by JacobBarthelmeh** [comment]
Thanks for the review of the code Todd. Will fix the unused functions, it slipped by from using GCC during development.

**Comment by toddouska** [comment]
Thanks for the changes, looks good!

**Pull request merged**